PR: https://github.com/odoo/odoo/pull/224987

From: 923f094ece27639481117843d3ecbe7a98636ca9
From: Bruno BOI
Date: 2026-08-25 07:17:01

Breaking data model changes scores: del:1 + add:0, change matches:
-    mail_post_method = fields.Selection(

Total Changes: 107

[IMP] mail,sms: simplify mail action server

Before this commit, server actions' mail feature were a bit confusing:
SMS sending with or without note did not make much sense
Message sending via email or message was not clearer and the user endud up
sending emails anyway.
Log note was absent or too complicated.

After this commit:
* mail_post action server behave like "Send a message"
  in a chatter the mail template retrieves default recipients,
  sends mail if needed and post in the chatter if the model
  is a mail.thread, otherwise it just sends emails
* added log_note state with only a html field that would be the body of
  the message
* sms state behaves like mail_post

task-4976562

closes odoo/odoo#224987

Related: odoo/enterprise#111930
Related: odoo/upgrade#9718
Signed-off-by: Lucas Perais (lpe) <lpe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mail/models/ir_actions_server.py
+++ b/addons/mail/models/ir_actions_server.py
@@ -1,6 +1,5 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
-
 from dateutil.relativedelta import relativedelta
 
 from odoo import _, api, fields, models
@@ -25,12 +24,14 @@ class IrActionsServer(models.Model):
         tracking=True,
         selection_add=[
             ('next_activity', 'Create Activity'),
-            ('mail_post', 'Send Email'),
+            ('mail_post', 'Send Message'),
+            ('log_note', 'Log Note'),
             ('followers', 'Add Followers'),
             ('remove_followers', 'Remove Followers'),
             ('code',),
         ],
         ondelete={'mail_post': 'cascade',
+                  'log_note': 'cascade',
                   'followers': 'cascade',
                   'remove_followers': 'cascade',
                   'next_activity': 'cascade',
@@ -62,22 +63,23 @@ class IrActionsServer(models.Model):
         default=lambda self: self.env.ref('mail.mt_comment').ids,
         readonly=False, store=True)
 
-    # Message Post / Email
+    # Message Post
     template_id = fields.Many2one(
-        'mail.template', 'Email Template',
+        'mail.template', 'Mail Template',
         domain="[('model_id', '=', model_id)]",
         compute='_compute_template_id',
         ondelete='set null', readonly=False, store=True,
     )
-    # Message post
+
     mail_post_autofollow = fields.Boolean(
         'Subscribe Recipients', compute='_compute_mail_post_autofollow',
         readonly=False, store=True)
-    mail_post_method = fields.Selection(
-        selection=[('email', 'Email'), ('comment', 'Message'), ('note', 'Note')],
-        string='Send Email As',
-        compute='_compute_mail_post_method',
-        readonly=False, store=True)
+
+    # Log note
+    log_note_note = fields.Html(
+        'Internal Note',
+        translate=True,
+        compute='_compute_log_note_note', readonly=False, store=True)
 
     # Next Activity: plan-based
     has_activity_plans = fields.Boolean(compute='_compute_has_activity_plans')
@@ -144,13 +146,27 @@ class IrActionsServer(models.Model):
     @api.depends('state')
     def _compute_available_model_ids(self):
         mail_thread_based = self.filtered(
-            lambda action: action.state in {'mail_post', 'followers', 'remove_followers', 'next_activity'}
+            lambda action: action.state in {'log_note', 'followers', 'remove_followers', 'next_activity'}
         )
+        to_super = self
+        # logging a note needs a mail.thread (its method _message_log)
+        # while sending an email with a template only needs a non-abstract model
+        # Do that sorting here
         if mail_thread_based:
             mail_models = self.env['ir.model'].search([('is_mail_thread', '=', True), ('transient', '=', False)])
             for action in mail_thread_based:
                 action.available_model_ids = mail_models.ids
-        super(IrActionsServer, self - mail_thread_based)._compute_available_model_ids()
+            to_super = to_super - mail_thread_based
+
+        need_non_abstract_model = self.filtered(lambda a: a.state == 'mail_post')
+        # Early mirror for mail_template._get_non_abstract_models_domain
+        if need_non_abstract_model:
+            models = self.env['ir.model'].search([('abstract', '=', False)])
+            for action in need_non_abstract_model:
+                action.available_model_ids = models.ids
+            to_super = to_super - need_non_abstract_model
+
+        super(IrActionsServer, to_super)._compute_available_model_ids()
 
     @api.depends('model_id', 'state')
     def _compute_template_id(self):
@@ -161,23 +177,15 @@ class IrActionsServer(models.Model):
         if to_reset:
             to_reset.template_id = False
 
-    @api.depends('state', 'mail_post_method')
+    @api.depends('state')
     def _compute_mail_post_autofollow(self):
-        to_reset = self.filtered(lambda act: act.state != 'mail_post' or act.mail_post_method == 'email')
-        if to_reset:
-            to_reset.mail_post_autofollow = False
-        other = self - to_reset
-        if other:
-            other.mail_post_autofollow = True
+        to_reset = self.filtered(lambda act: act.state != 'mail_post')
+        to_reset.mail_post_autofollow = False
 
     @api.depends('state')
-    def _compute_mail_post_method(self):
-        to_reset = self.filtered(lambda act: act.state != 'mail_post')
-        if to_reset:
-            to_reset.mail_post_method = False
-        other = self - to_reset
-        if other:
-            other.mail_post_method = 'comment'
+    def _compute_log_note_note(self):
+        to_reset = self.filtered(lambda act: act.state != 'log_note')
+        to_reset.log_note_note = False
 
     @api.depends('model_id', 'state')
     def _compute_followers_type(self):
@@ -281,7 +289,8 @@ class IrActionsServer(models.Model):
 
     @api.model
     def _warning_depends(self):
-        return super()._warning_depends() + [
+        return [
+            *super()._warning_depends(),
             'activity_date_deadline_range',
             'model_id',
             'template_id',
@@ -301,12 +310,11 @@ class IrActionsServer(models.Model):
         if self.state == 'mail_post' and self.template_id and self.template_id.model_id != self.model_id:
             warnings.append(_("Mail template model of $(action_name)s does not match action model.", action_name=self.name))
 
-        if self.state in {'mail_post', 'followers', 'remove_followers', 'next_activity'} and self.model_id.transient:
-            warnings.append(_("This action cannot be done on transient models."))
+        if self.state in {'mail_post', 'log_note', 'followers', 'remove_followers', 'next_activity'} and (self.model_id.transient or self.model_id.abstract):
+            warnings.append(_("This action cannot be done on transient or abstract models."))
 
         if (
-            (self.state in {"followers", "remove_followers"}
-            or (self.state == "mail_post" and self.mail_post_method != "email"))
+            self.state in {"followers", "remove_followers", "log_note"}
             and not self.model_id.is_mail_thread
         ):
             warnings.append(_("This action can only be done on a mail thread models"))
@@ -385,11 +393,28 @@ class IrActionsServer(models.Model):
                     return True
         return False
 
+    def _run_action_log_note_multi(self, eval_context=None):
+        if (not self.env.context.get('active_ids') and not self.env.context.get('active_id')) or self._is_recompute():
+            return False
+        res_ids = self.env.context.get('active_ids', [self.env.context.get('active_id')])
+
+        # Clean context from default_type to avoid making attachment
+        # with wrong values in subsequent operations
+        cleaned_ctx = dict(self.env.context)
+        cleaned_ctx.pop('default_type', None)
+        cleaned_ctx.pop('default_parent_id', None)
+
+        records = self.env[self.model_name].with_context(cleaned_ctx).browse(res_ids)
+        bodies = {rec.id: self.log_note_note for rec in records}
+        records._message_log_batch(bodies, message_type="notification")
+        return False
+
     def _run_action_mail_post_multi(self, eval_context=None):
-        # TDE CLEANME: when going to new api with server action, remove action
         if not self.template_id or (not self.env.context.get('active_ids') and not self.env.context.get('active_id')) or self._is_recompute():
             return False
         res_ids = self.env.context.get('active_ids', [self.env.context.get('active_id')])
+        if res_ids and self.model_id != self.template_id.model_id:
+            raise ValueError(self.env._("The model of the mail template is different from the one on the action server"))
 
         # Clean context from default_type to avoid making attachment
         # with wrong values in subsequent operations
@@ -398,27 +423,17 @@ class IrActionsServer(models.Model):
         cleaned_ctx.pop('default_parent_id', None)
         cleaned_ctx['mail_post_autofollow_author_skip'] = True  # do not subscribe random people to records
         cleaned_ctx['mail_post_autofollow'] = self.mail_post_autofollow
-
-        if self.mail_post_method in ('comment', 'note'):
+        if isinstance(self.env[self.model_name], self.env.registry.get("mail.thread")):
             records = self.env[self.model_name].with_context(cleaned_ctx).browse(res_ids)
-            message_type = 'auto_comment' if self.state == 'mail_post' else 'notification'
-            if self.mail_post_method == 'comment':
-                subtype_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_comment')
-            else:
-                subtype_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
+            subtype_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_comment')
             records.message_post_with_source(
                 self.template_id,
-                message_type=message_type,
+                message_type='auto_comment',
                 subtype_id=subtype_id,
             )
         else:
             template = self.template_id.with_context(cleaned_ctx)
-            for res_id in res_ids:
-                template.send_mail(
-                    res_id,
-                    force_send=False,
-                    raise_exception=False
-                )
+            template.send_mail_batch(res_ids, force_send=False, raise_exception=False)
         return False
 
     def _run_action_next_activity_multi(self, eval_context=None):
